processing/src/main/java/io/druid/query/topn/PooledTopNAlgorithm.java

```diff
@@ -194,16 +207,81 @@ protected void scanAndAggregate(
   )
   {
     final Cursor cursor = params.getCursor();
-    if (specializeGeneric1AggPooledTopN && theAggregators.length == 1) {
-      scanAndAggregateGeneric1Agg(params, positions, theAggregators[0], cursor);
-    } else if (specializeGeneric2AggPooledTopN && theAggregators.length == 2) {
+    if (theAggregators.length == 1) {
+      BufferAggregator aggregator = theAggregators[0];
+      if (cursor instanceof HistoricalCursor && aggregator instanceof SimpleDoubleBufferAggregator) {
+        if (specializeHistoricalSingleValueDimSelector1SimpleDoubleAggPooledTopN &&
```

Comment:
To be understood how to cover more cases without turning `PooledTopNAlgorithm` into a scary wall of if-else text like this.